queries/kotlin.scm

```diff
@@ -0,0 +1,627 @@
+  (#not-type? @statement statements)
+)
+
+(class_body
+  (_) @statement
+)
+
+(enum_class_body
+  (_) @statement
+)
+
+(class_declaration
+  (type_identifier) @name @className
+) @class @_.domain
+
+(object_declaration
+  (type_identifier) @name @className
+) @class @_.domain
+
+(companion_object
+  (type_identifier) @name @className
+) @class @_.domain
+
+(function_declaration
+  (simple_identifier) @name @functionName
+) @namedFunction @_.domain
+(secondary_constructor) @namedFunction
+
+(if_expression) @ifStatement
+
+;;
+;; Literals and comments
+;;
+
+(anonymous_function) @anonymousFunction
+
+(lambda_literal) @anonymousFunction
+
+(string_literal) @string @textFragment
+
+[
+  (line_comment)
+  (multiline_comment)
+] @comment @textFragment
+
+;;
+;; Branches and conditions
+;;
+
+(when_entry) @branch
+
+(when_entry
+  [
+    (when_condition)
+    "else"
+  ] @condition
+  (#allow-multiple! @condition)
+) @_.domain
+
+(when_expression) @branch.iteration @condition.iteration
+
+(if_expression
+  "if"
+  .
+  "("
+  .
+  (_) @condition
+  .
+  ")"
+) @_.domain
+
+;; If branch
+(if_expression
+  "if" @branch.start @branch.removal.start
+  .
+  "("
+  .
+  (_)
+  .
+  ")"
+  .
+  (control_structure_body) @branch.end @branch.removal.end
+  (
+    "else"
+    (control_structure_body) @branch.removal.end.startOf
+  )
+?
+)
+
+;; Else-if branch
+(if_expression
+  "else" @branch.start @condition.domain.start
+  (control_structure_body
+    (if_expression
+      "if"
+      .
+      "("
+      .
+      (_) @condition
+      .
+      ")"
+      .
+      (control_structure_body) @branch.end @condition.domain.end
+    )
+  )
+)
+
+;; Else branch
+(if_expression
+  "else" @branch.start
+  (control_structure_body) @branch.end
+)
+
+(while_statement
+  "while"
+  .
+  "("
+  .
+  (_) @condition
+  .
+  ")"
+) @_.domain
+
+(do_while_statement
+  "("
+  .
+  (_) @condition
+  .
+  ")"
+  .
+) @_.domain
+
+(when_expression
+  (when_subject) @private.switchStatementSubject
+  (#child-range! @private.switchStatementSubject 0 -1 true true)
+) @_.domain
+
+;;
+;; Name, value, type, and key
+;;
+
+(type_alias
+  "typealias"
+  .
+  (_) @name.start
+  (_)? @name.end
+  .
+  "="
+  .
+  (_) @value.start @type.start
+) @value.end.endOf @type.end.endOf @_.domain
+
+(class_parameter
+  (simple_identifier) @name
+) @_.domain
+
+(class_parameter
+  ":"
+  .
+  (_) @type.start
+  (_)? @type.end
+  .
+  "="
+  (_) @value
+) @_.domain
+
+;; Known issue: this won't work with multiple-node types.
+(class_parameter
+  ":"
+  .
+  (_) @type
+  .
+) @_.domain
+
+;; Function declarations with type constraints
+(function_declaration
+  ":"
+  .
+  (_) @type.start
+  (_)? @type.end
+  .
+  (type_constraints)
+) @_.domain
+
+;; Function declarations with no type constraints but with body
+(
+  (function_declaration
+    ":"
+    .
+    (_) @type
+    .
+    (function_body)
+  ) @_.domain
+)
+(
+  (function_declaration
+    ":"
+    .
+    (_) @type.start
+    (_) @type.end
```

Comment:
If I could just do `(_)? @type.end` I wouldn't need this redundant query, but Cursorless complains when it's missing because I refer to it in a predicate below. I think Cursorless could safely ignore constraints when they don't appear in cases such as this where they are tied to an optional match. My reasoning is that if the predicate must be confirmed, then it shouldn't be an optional match. Should I file a bug for this? There are several places throughout this file where I could simplify my code if this were possible.